src/session.c

```diff
@@ -178,56 +180,129 @@ static P11PROV_SESSION *p11prov_session_ref(P11PROV_SESSION *session)
+ * Returns false on an error or if max_wait is exceeded.
+ */
+static bool cyclewait_with_timeout(uint64_t max_wait, uint64_t interval,
+                                   uint64_t *start_time)
+{
+#define NANOS_SEC 1000000000
+    uint64_t current_time;
+    struct timespec ts;
+    int ret;
+
+    ret = clock_gettime(CLOCK_MONOTONIC, &ts);
+    if (ret != 0) {
+        return false;
+    }
+
+    current_time = ts.tv_sec * NANOS_SEC + ts.tv_nsec;
+    if (*start_time == 0) {
+        *start_time = current_time;
+    } else {
+        if (current_time > *start_time + max_wait) {
+            return false;
+        }
+    }
+
+    ts.tv_sec = interval / NANOS_SEC;
+    ts.tv_nsec = interval % NANOS_SEC;
+    ret = nanosleep(&ts, NULL);
+    if (ret != 0 && ret != EINTR) {
+        return false;
+    }
+
+    return true;
+}
+
+/* in nanoseconds, 1 seconds */
+#define MAX_WAIT 1000000000
+/* sleep interval, 50 microseconds (max 20 attempts) */
+#define SLEEP 50000
+static CK_RV internal_session_open(P11PROV_SESSION_POOL *p, CK_SLOT_ID slot,
+                                   CK_SESSION_HANDLE *session)
 {
-    P11PROV_SESSION_POOL *p = session->pool;
-    CK_ULONG cur_sessions;
     CK_FUNCTION_LIST *f;
+    bool wait_ok = true;
+    CK_ULONG cs = 0;
+    uint64_t startime = 0;
     CK_RV ret;
 
-    f = p11prov_ctx_fns(session->provctx);
+    f = p11prov_ctx_fns(p->provctx);
     if (f == NULL) {
         return CKR_FUNCTION_FAILED;
     }
 
-    cur_sessions = __atomic_add_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
-    if (cur_sessions > p->max_sessions) {
-        ret = CKR_SESSION_COUNT;
-        P11PROV_raise(session->provctx, ret, "Max sessions limit");
-        goto done;
+    while (wait_ok) {
+        if (cs == 0) {
+            cs = __atomic_add_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+            if (cs > p->max_sessions) {
+                (void)__atomic_sub_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+                ret = CKR_SESSION_COUNT;
+                cs = 0;
+                wait_ok = cyclewait_with_timeout(MAX_WAIT, SLEEP, &startime);
+                continue;
+            }
+        }
+
+        wait_ok = false;
+        ret = f->C_OpenSession(slot, CKF_SERIAL_SESSION, NULL, NULL, session);
+        if (ret == CKR_SESSION_COUNT) {
+            wait_ok = cyclewait_with_timeout(MAX_WAIT, SLEEP, &startime);
+        }
     }
 
-    ret = f->C_OpenSession(session->slotid, CKF_SERIAL_SESSION, NULL, NULL,
-                           &session->session);
+    if (ret != CKR_OK && cs != 0) {
+        (void)__atomic_sub_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+    }
+    return ret;
+}
+
+static CK_RV p11prov_session_open(P11PROV_SESSION *session, bool login,
+                                  CK_UTF8CHAR_PTR pin, CK_ULONG pinlen)
+{
+    P11PROV_SESSION_POOL *p = session->pool;
+    CK_RV ret;
+
+    ret = internal_session_open(p, session->slotid, &session->session);
```

Comment:
I think this is wrong place for the waiting. It is possible that other threads exhausted the internal P11PROV_SESSION objects, and we don't even get here.
The right place for synchronizing/waiting would be `p11prov_get_session`.

Reply:
Uhmm, I guess we can avoid the check in session_new then, however should we have an upper limit to p11prov_sessions too ?
I do not think it make sense to have too many of them way beyond the max number of possible token sessions. they will give a lot of "free" sessions that just keep failing with a timeout later if no sessions are being released, or the token somehow malfunctioned and decided to not give out sessions anymore.